src/backend_cypressfx3/sw/backend_cypressfx3.c

```diff
@@ -454,6 +455,26 @@ int gb_cypressfx3_close(struct glip_ctx *ctx)
         return -1;
     }
 
+    /* ensure the write buffer is empty before closing the connection */
+    struct glip_backend_ctx* bctx = ctx->backend_ctx;
+    size_t write_buf_fill_level = cbuf_fill_level(bctx->write_buf);
+    if (write_buf_fill_level > 0) {
+		struct timespec ts;
+		/* get current time */
+		timespec_get(&ts, TIME_UTC);
+		/* add 100ms as timeout */
+		ts.tv_nsec += 100000000;
+		int res = 0;
+		while (write_buf_fill_level > 0 && res == 0) {
+			res = cbuf_timedwait_for_level_change(bctx->write_buf,
+					write_buf_fill_level, &ts);
+			write_buf_fill_level = cbuf_fill_level(bctx->write_buf);
+		}
+		/* needed because otherwise the last message is sometimes not written */
+		/* also, without it the program sometimes stalls at 'libusb_close()' */
```

Comment:
- Actually, the first comment should read "Ensure that data which has been written using glip_write() is written out to the USB device.", or not?
- The second comment isn't true? That's avoided by the setcancelstate which we inserted in the write thread.